### PR TITLE
60468 handle BTSSS token error

### PR DESCRIPTION
**Pull request by randomsync**
## Summary
This PR adds error handling for BTSSS authentication call to `token` endpoint. In case of error, an sms is sent to the user.

team: check-in

## Related issue(s)
- department-of-veterans-affairs/va.gov-team#60468

## Testing done
- Local testing
- rspecs

## What areas of the site does it impact?
* check_in module

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### CODE REVIEW

File: modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb
Comment (javierfeldman) [inline modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:44]: Should it send a notification if the call fails?

If it should just log and exit, I'd add a `return unless claims_resp.present?` above like 41.

File: modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb
Comment (javierfeldman) [inline modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:41]: This code looks similar to what the `handle_response` code. Maybe the same method could handle both values and return both.

```ruby
      claim_number, template_id = handle_response(claims_resp:)
```

File: modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb
Comment (javierfeldman) [inline modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:37]: Should the code continue after the API call fails? If no, you should add a `return unless claim_resp` after this begin/rescue block.

File: modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb
Comment (javierfeldman) [inline modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:41]: This looks similar to what is being done within `handle_response` so maybe the same method could handle both value.

`claim_number, template_id = handle_response(claims_resp:)`

**Comment by ryan-mcneil** [comment]
@randomsync Hi! Just wanted to make sure you weren't waiting on us. We are slowly rolling out some new policies in which we're asking for devs to get approvals from someone on their team before coming to us for the final approval. It's impossible for us to have context and understanding of business requirements for every PR, so it's much easier if we get that first 👍 from someone closer to the action. We'll keep an eye out for green check marks under `Reviewers`. Thank you!

**Comment by randomsync** [comment]
@ryan-mcneil thanks, yes that makes sense. The engineer from my team was OOO and she just got back. I'll review this with her and get the approvals.

**Review by kanchanasuriya — DISMISSED** [review]
LGTM

File: modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb
Comment (OrganicCat) [inline modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:39]: If you `return nil` rather than setting claims_resp to nil, that would solve future issues of handling nil values in claims_resp, wouldn't it? It prevents the rest of the code from running. Or if you want the rest of the code to run anyway, setup another function perhaps to handle the nil value path vs the good path. Some ideas if you wanted to short-circuit the `claims_resp` as `nil`

**Review by OrganicCat — COMMENTED** [review]
If this wouldn't work, no problem, tossing the idea out there

File: modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb
Comment (lee-delarm6) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb]: Woops, this is Lee, sorry I commented with the wrong account

**Review by lee-delarm6 — DISMISSED** [review]
Looks good!

**Review by OrganicCat — APPROVED** [review]

**Comment by jeremy6d** [comment]
Please do us on support a solid and remove filler text from the PR in the future -- we want to be able to focus on the gist of the issue and not have to filter out text like:

> Note: Delete the description statements, complete each step. None are optional, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

**Review by jeremy6d — DISMISSED** [review]
I think it's safest for the platform if exception handling is dialed in to the specific exception you are expecting to handle. Reach out to me on slack if I can provide more context.

File: modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb
Comment (randomsync) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb]: The `submit_claim` method above calls:
- VEIS token endpoint if the token isn't already present in cache, and
- `submitclaim` endpoint 

The method already handles any exception from `submitclaim` call and returns a handled response. The token call however raises an exception. The goal of this rescue block is to handle that exception and log it, emit a metric and send a notification. I've added the exception message to the logger so we can monitor it. Even though we're sure it'll be a `BackendServiceException` in this case, I'm leaving it generic just in a remote chance someone adds a more specific exception mapping to the client later. thoughts?

Comment (randomsync) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb]: addressed in 80e0cc1e00f2ba3f19b2067d0696a5771eab14ec

File: modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb
Comment (kanchanasuriya) [inline modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: We would miss the notification if there are any changes with exception mapping in future.  Wondering if we can make it generic like before as we have an exception message logged.

Comment (randomsync) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: please see the conversation above. it still gets cast into a `BackendServiceException`

Comment (kanchanasuriya) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: Yeah, that's the concern.  For example, if there is any `InternalServiceException`, the rescue block will not handle & a notification will not be sent.

Comment (randomsync) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: why would there be `InternalServiceException`? if there's a service exception related to `token` call, wouldn't that get thrown as `BackendServiceException`?

Comment (OrganicCat) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: If we wanted to be thorough, couldn't we just handle the one expected exception call with the first rescue block, and add a second rescue block for non-standard catch-all errors?

Comment (lee-delarm6) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: If we needed to be thorough, can't we handle the expected exception, and then add a second rescue block to catch any unexpected exceptions? This would seem to me to handle the best and worst use cases

Comment (kanchanasuriya) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: > why would there be `InternalServiceException`? if there's a service exception related to `token` call, wouldn't that get thrown as `BackendServiceException`?

Just sighted a worst-case scenario if there is an issue in mapping the exception to BackendServiceException

Comment (kanchanasuriya) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: > If we needed to be thorough, can't we handle the expected exception, and then add a second rescue block to catch any unexpected exceptions? This would seem to me to handle the best and worst use cases

Ideally, yes, this approach is thorough.  But we do not see any difference in both the rescue block.  Guess, we can keep it simple.

Comment (randomsync) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: > > why would there be `InternalServiceException`? if there's a service exception related to `token` call, wouldn't that get thrown as `BackendServiceException`?
> 
> Just sighted a worst-case scenario if there is an issue in mapping the exception to BackendServiceException

I'm not sure it applies here: "issue in mapping the exception"...are you able to provide a specific code path where that would happen?

Comment (randomsync) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: As an option, I can catch the `StandardError` which is the base class that most of the Exception classes extend from. Would that satisfy both sides' concerns?

Comment (kanchanasuriya) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: Yes, that would work also.

Comment (randomsync) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: 😆 
image

Comment (kanchanasuriya) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: I feel this [commit](https://github.com/department-of-veterans-affairs/vets-api/pull/13085/commits/6ae7d4e30ab58fc8e895f9cfe68057692403cf43) would handle all the exceptions with relevant exception message in the log.

Comment (randomsync) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: @kanchanasuriya I have gone through the code and I don't see any path where anything other than a BackendServiceException would be raised. Moreover, we have sentry logging both on the `token` method and `submit_claim` method that would alert us of any exceptions.

Comment (kanchanasuriya) [reply modules/check_in/app/workers/check_in/travel_claim_submission_worker.rb:38]: Thanks for the additional effort with code dive deep.  As discussed offline, approving as we have logs to alert the team on any exceptions.

**Review by kanchanasuriya — APPROVED** [review]
LGTM

**Review by rmtolmach — APPROVED** [review]
LGTM

**Comment by rmtolmach** [comment]
GitGuardian is failing b/c of https://github.com/department-of-veterans-affairs/vets-api/pull/13100 and can be ignored in this PR.

**Pull request merged**
